File: QSort/Cpp/QSort.cpp
#include <iostream>
#include <string>
#include <random>
#include <vector>

using namespace std;

vector<int> Swap(vector<int> ar,int i,int j) {
	int tmp = ar[i];
	ar[i]=ar[j];
	ar[j]=tmp;
	return ar;
}

void PrintAr(vector<int> ar,int n) {
	cout << "[ " << flush;
	for (int i=0; i<n; i++) {
		cout << ar[i];
		if (i != n-1)
			cout << ", ";
	}
	cout << " ]" << endl;
}

vector<int> QSort(vector<int> ar, int left, int right) {
	int i = left;
	int j = right;
	int pivot = ar[left];
	while (1) {
		while (ar[i] < pivot)
			i++;
		while (pivot < ar[j])
			j--;
		if (i >= j)
			break;
		ar=Swap(ar,i,j);
		i++;
		j--;
	}
	if (left < i-1)
		ar=QSort(ar, left, i-1);
	if (j+1 < right)
		ar=QSort(ar, j+1, right);
	return ar;
}


int main() {
	string str;
	int n;
	
	// Input
	cout << "n = ";
	getline(cin,str);
	try {
		n = stoi(str);
	}
	catch(exception &e) {
		n = 0;
	}
	if (n < 10) {
		n = 10;
	}
	cout << endl;
	
	// Initialize
	vector<int> ar(n);
	for (int i = 0; i < n; i++) {
		ar[i] = i+1;
	}
	
	// Shuffle
	random_device rd;
	mt19937 mt(rd());
	uniform_int_distribution<int> rand(0, n-1);
	for (int i = n; i > 1; i--) {
        ar=Swap(ar,i-1,rand(mt)%i);
    }
    
    PrintAr(ar,n);
    ar = QSort(ar,0,n-1);
    PrintAr(ar,n);
}

File: QSort/Python3/QSort.py
import random

def Swap(intlist,a,b):
	buf = intlist[a]
	intlist[a]=intlist[b]
	intlist[b]=buf
	return

def Shuffle(intlist,n):
	for i in intlist:
		j = n - i
		k = random.randrange(n) % (j+1)
		Swap(intlist,j,k)
		
def QSort(intlist,left,right):
	i = left
	j = right
	pivot = intlist[left]
	while True:
		while intlist[i] < pivot:
			i+=1
		while pivot < intlist[j]:
			j-=1
		if i >= j:
			break
		Swap(intlist,i,j)
		i+=1
		j-=1
	if left < i-1:
		QSort(intlist, left, i-1)
	if j+1 < right:
		QSort(intlist, j+1, right)

print('n = ', end = '')
try:
	n = int(input())
except:
	n = 10
if n < 10:
	n = 10
print('')
intlist = list(range(n))
for i in intlist:
	intlist[i]=i+1

print('Initialize : ')
print(intlist)
print('')
Shuffle(intlist,n)
print('Shuffle : ')
print(intlist)
print('')
QSort(intlist,0,n-1)
print('Sort : ')
print(intlist)
print('')